File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[]}
     */
    productExceptSelf(nums) {
        let result = [];
        let leftProducts = [];
        let rightProducts = [];
        let leftProduct = 1;
        let rightProduct = 1;
        
        for (let i = 0; i < nums.length; i++) {
            leftProducts[i] = leftProduct;
            leftProduct *= nums[i];
        }

        for (let i = nums.length - 1; i >= 0; i--) {
            rightProducts[i] = rightProduct;
            rightProduct *= nums[i];
        }
        
        for (let i = 0; i < nums.length; i++) {
            result[i] = leftProducts[i] * rightProducts[i];
        }

        return result;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        pref = [1] * len(nums)
        suf = [1] * len(nums)
        
        for i in range(1, len(nums), 1):
            pref[i] = pref[i-1] * nums[i-1]
        
        for i in range (len(nums) - 2, -1, -1):
            suf[i] = suf[i + 1] * nums[i+1]
        
        res = [1] * len(nums)
        for i in range(len(nums)):
            res[i] = pref[i] * suf[i]

        return res